WIP: custom Unicode normalization for Julia identifiers

stevengj opened the pull request:
As discussed in #14751, #5903, and JuliaLang/utf8proc#11, this PR canonicalizes Unicode identifiers/symbols in Julia via a custom normalization, in addition to NFC normalization.  That is, we treat certain codepoints as equivalent in identifiers.  (Requires a new feature in the soon-to-be-released utf8proc 2.1.)

I don't think we want to go too crazy on "confusable" characters here—we made a conscious choice in #5434 to use NFC normalization, not NFKC normalization (ala Python 3), in order to preserve useful mathematical distinctions.  I propose the following criteria for inclusion in our custom normalization, to be **applied conservatively when real problems arise** in the wild:

* Both characters are easy to type by common input methods.
* The two characters are so close in appearance that it makes no sense to use them for distinct identifiers in the same scope, and people might reasonably type one when the other is intended.

Currently, this PR canonicalizes ɛ (U+025B latin small letter open e) to ε (U+03B5 greek small letter epsilon) as discussed in #14751, and canonicalizes µ (U+00B5 micro) to μ (U+03BC greek small letter mu) as discussed in #5434.   I think it would also be good to canonicalize fullwidth to halfwidth identifiers as discussed in #5903 (one of the primary reasons Python 3 went with NFKC).

cc: @StefanKarpinski, @jiahao

stevengj committed:
implement custom Julia Unicode normalization for confusable characters in identifiers

stevengj committed:
whoops

stevengj committed:
separated julia_charmap into its own file to make it easier to update

iamed2 commented:
Could the exact additions to NFC be documented in the manual? And maybe another method of `normalize_string`?

stevengj commented:
Yes, we should definitely document it.  You automatically get the "Julia" normalization just by converting a string to a symbol, so I'm not 100% sure we need another API.

iamed2 commented:
Oh, I didn't know that!

stevengj committed:
normalize fullwidth -> halfwidth in identifiers, ala NFKC

stevengj committed:
make \varepsilon complete to ε (u+03b5), fixes #14751

stevengj committed:
docs for canonicalization

stevengj committed:
normalize fullwidth characters during parsing (fixes #5903)

stevengj commented:
Added documentation.

I also added normalization of fullwidth to halfwidth characters (using a subset of NFKC), and I hooked this in at parse time so that things like `b＝3:5` now work using the fullwidth `＝`  (#5903).  

@jiahao, in particular, I applied the NFKC normalization to codepoings in `[0xff01:0xff5e; 0x3000; 0xffe0:0xffee]`, which seems to be what [Wikipedia says are the fullwidth forms](https://en.wikipedia.org/wiki/Halfwidth_and_fullwidth_forms) of ASCII and other symbols.  I would appreciate your expertise here regarding anything else that might be needed.

stevengj committed:
typo

stevengj commented:
@iamed2, I take that back: `Symbol(s::String)` does *not* normalize `s`.  However, `parse(s::String)` does if `s` is valid identifier.

pabloferz commented:
Another pair of characters that I believe follow the proposed criteria are `ĸ` (U+0138 latin small letter kra) and `κ` (U+03BA greek small letter kappa). The first one is easy to type in some keyboards.

stevengj committed:
tests

stevengj commented:
Hmm, not quite working yet.  For one thing, I was too aggressive in normalizing every character read by the parser.  We definitely don't want to normalize characters that appear in string and character literals.

stevengj commented:
@pabloferz, in what circumstances are you likely to type a "kra" when kappa is intended?

stevengj commented:
Another question for @jiahao: should we only normalize fullwidth identifiers and operators, or should we also normalize fullwidth quotation marks and @ and similar syntactically important symbols?  I would think that the answer is yes, as otherwise when typing Julia code in Pinyin mode you'll continually have to switch back to English.

pabloferz commented:
In a bunch of Linux keyboard layouts typing Right Alt + m gives 'micro' while Right Alt + k gives 'kra'.

stevengj commented:
Ah, then kra -> kappa indeed makes sense.

stevengj committed:
be more cautious about normalizing chars when parsing, so as not to normalize string literals

stevengj committed:
test fullwidth numeric literals and parens

stevengj commented:
[Kra](https://en.wikipedia.org/wiki/Kra_(letter)) seems like a weird case, because apparently it is more like a type of "q" than a type of "k".  It seems like it would be like typing ß (german ss) for β (beta).   I'm hesitant to do this normalization unless we have more evidence that it is something people are really likely to want.

stevengj commented:
Okay, I fixed the earlier problems.

Now it does fullwidth->ASCII normalization on operators and identifiers, but also on numeric literals and punctuation (parens, brackets, @, commas, #comments)... basically everything *except* string and character literals.  (Quotation marks still need to be ASCII, too.)  So basically you can type all your Julia code except for string/character literals in Pinyin input mode, without switching back and forth.

nalimilan commented:
It's kind of weird that Linux layouts offer Kra rather than Kappa. I guess that's because it's considered as latin? Yet `Right Alt  + m` gives Mu, which is a greek letter. Maybe we should file a bug.

tkelman commented on deps/utf8proc.version:
```diff
@@ -1,2 +1,2 @@
-UTF8PROC_BRANCH=v2.0.2
-UTF8PROC_SHA1=e3a5ed7b8bb5d0c6bb313d3e1f4d072c04113c4b
+UTF8PROC_BRANCH=master
```
this will need to be a release tag, and update checksums when ready

tkelman commented on doc/manual/variables.rst:
```diff
@@ -134,6 +134,14 @@ statements:
+some Asian languages.  The Unicode characters ``ɛ`` (U+025B: Latin small letter open e)
+and ``µ`` (U+00B5: micro sign) are treated as equivalent to the corresponding
+Greek letters, because the former are easily accessible via some input methods.
+Different ways of entering Unicode combining are treated as equivalent
```
combining... characters?

stevengj committed:
typo/clarification

stevengj commented:
It would be nice to get some feedback on the main decision point here: whether it is desirable to be able to write Julia code (except for literal strings) as either/both ASCII and fullwidth and have them be parsed as equivalent.

I talked briefly with @jiahao about it.  On the one hand, it would make it much more convenient to use East Asian-language input modes when writing Julia code.  Most Asian-language programmers have learned not to do this because no other programming language allows it. On the other hand, fullwidth code looks a little weird, especially to programmers coming from other languages, and we have to decide whether that diversity in appearance is something we want to encourage.

stevengj committed:
update to utf8proc-2.1

stevengj commented:
Rebased.   Any chance of a decision on this, or is this postponed to the next release?

tkelman commented on deps/utf8proc.version:
```diff
@@ -1,2 +1,2 @@
-UTF8PROC_BRANCH=v2.0.2
-UTF8PROC_SHA1=e3a5ed7b8bb5d0c6bb313d3e1f4d072c04113c4b
+UTF8PROC_BRANCH=v2.1
+UTF8PROC_SHA1=40e605959eb5cb90b2587fa88e3b661558fbc55a
```
update the checksums

stevengj replied:
thanks, fixed

tkelman commented on test/core.jl:
```diff
@@ -3364,6 +3364,25 @@ typealias PossiblyInvalidUnion{T} Union{T,Int}
 @test Symbol("x") === Symbol("x")
 @test split(string(gensym("abc")),'#')[3] == "abc"
 
+# normalization of Unicode symbols (#19464)
```
these should probably be in test/parse, not core

stevengj replied:
done.

stevengj committed:
checksum for utf8proc 2.1

stevengj committed:
moved symbol-normalization test from test/core to test/parse

tkelman commented:
The fullwidth part of this seems a bit questionable to me.

stevengj commented:
I can always put the fullwidth part (which is mostly relevant to speakers Chinese and Japanese languages, it seems) in a separate PR if we want to keep the ε and μ parts.  But the motivations seem rather similar to me: common input methods allow you to type the "same" character as different codepoints.

stevengj commented:
Rebased again.  @StefanKarpinski, do you have an opinion on this?

It would be useful to get some feedback from native speakers of Chinese or Japanese on the desirability of fullwidth Julia code, since it would presumably only be used in contexts where one is doing Pinyin input or similar.  cc @jiahao, @yuyichao, @lindahua, @chezou

chezou commented:
As a native Japanese speaker, IMHO it is too much to convert full-width into half-width, especially for operators like equal.

Japanese usually use ASCII mode for coding and switch to Japanese mode if we input Japanese comments or Japanese string. It also might include some other problem, such as wave dash problem [1](https://en.wikipedia.org/wiki/Tilde#Unicode_and_Shift_JIS_encoding_of_wave_dash) [2](https://en.wikipedia.org/wiki/Japanese_punctuation#Wave_dash).

We don't mistake writing of full-dash and half-dash, but might mistake reading of full-dash and half-dash.

stevengj committed:
Revert "be more cautious about normalizing chars when parsing, so as not to normalize string literals"

This reverts commit 81033fa0a0f11f9e42907e3348c64b95f070cac4.

stevengj committed:
Revert "normalize fullwidth characters during parsing (fixes #5903)"

This reverts commit cf61972a68bc76b4f2fdbebc5a9fb3e5c9619338.

stevengj committed:
remove more references to fullwidth normalization

stevengj commented:
I've removed the fullwidth->halfwidth normalization from this PR, since it is controversial.  We can also add it in another PR.  (I've saved the fullwidth changes in [another branch](https://github.com/stevengj/julia/tree/id_norm_fullwidth).)

stevengj committed:
rm fullwidth identifier normalization

stevengj commented:
Putting the 0.6 milestone on to make sure we make a decision about this.  Currently, the PR just normalizes the latin ε and micro μ.

StefanKarpinski commented:
In favor. Needs squashing, but otherwise seems good to go? (Failure is the usual on 32-bit.)

stevengj commented:
Yup, should be good to squash and merge.

tkelman committed:
Merge branch 'master' into id_norm

tkelman commented on src/flisp/julia_extensions.c:
```diff
@@ -152,22 +152,55 @@ value_t fl_julia_identifier_start_char(fl_context_t *fl_ctx, value_t *args, uint
     ssize_t result;
     size_t newlen;
-    result = utf8proc_decompose((uint8_t*) s, 0, NULL, 0, (utf8proc_option_t)options);
+    result = utf8proc_decompose_custom((uint8_t*) s, 0, NULL, 0, (utf8proc_option_t)options,
```
can we put this behind a utf8proc-version conditional so it's still possible to build against older utf8proc version, just without having this feature? cc @nalimilan

nalimilan replied:
It could be confusing to have Julia builds with normalization and others without. Personally, for RPM packages, I'll also provide a new version of utf8proc where needed (even if it would of course be simpler not to have to do this).

tkelman replied:
In that case we should document 2.1 as the minimum allowable version.

stevengj replied:
Fixed in 2f223b71b9ae394a76da187ff96f9a945dc9f2a5

Keno commented:
It would be nice to have this as an option to `normalize_string`.

stevengj commented:
@Keno, you can always just call `parse` on an identifier string...

Keno commented:
Well a) not if you're trying to replace the parser and b) that seems like a pretty roundabout way of doing it.
